### ISSUE TITLE
Nil pointer dereference error (js tracing)

### ISSUE THREAD

**Issue by holiman** [ethereum/go-ethereum#17355]
@gballet found a way to break the mon machine, via some sort of js tracing. 
```
Aug 08 15:14:34 mon02.ethdevops.io geth-devel: INFO [08-08|13:14:34.432] Imported new chain segment               blocks=1 txs=11  mgas=0.542  elapsed=26.086ms  mgasps=20.771  number=6110647 hash=b08ae7…3ffa7d cache=536.11mB 
Aug 08 15:14:40 mon02.ethdevops.io geth-devel: INFO [08-08|13:14:40.075] Preparing state for chain trace          block=5701347 start=5701347 elapsed=418.674µs 
Aug 08 15:14:43 mon02.ethdevops.io geth-devel: WARN [08-08|13:14:43.857] Tracing failed                           hash=a858b5…8023c6 block=5701348 err="execution timeout" 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: WARN [08-08|13:14:44.005] Chain tracing aborted                    start=5701346 end=5727531 abort=5701348 transactions=132 elapsed=3.930023922s 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: panic: runtime error: invalid memory address or nil pointer dereference 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x69e99e] 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: goroutine 6917801 [running]: 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: github.com/ethereum/go-ethereum/trie.(*Database).reference(...) 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: 	/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/trie/database.go:423 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: github.com/ethereum/go-ethereum/trie.(*Database).Reference(0xc446601a40, 0x55feb18218051d0a, 0x7333d0b2c1db49b7, 0x8b7e28c99111e342, 0x9d0f705446bb1b66, 0x0, 0x0, 0x0, 0x0) 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: 	/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/trie/database.go:412 +0x18e 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: github.com/ethereum/go-ethereum/eth.(*PrivateDebugAPI).traceChain.func2(0xc4313a9c20, 0xc42355ff90, 0xc47fcfcb40, 0xc4b6129290, 0xbed2d91004796d8e, 0x60734224bb8b, 0x1b5a900, 0xc4b7b65260, 0xc4b429a540, 0x56fee3, ...) 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: 	/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/eth/api_tracer.go:295 +0x894 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: created by github.com/ethereum/go-ethereum/eth.(*PrivateDebugAPI).traceChain 
Aug 08 15:14:44 mon02.ethdevops.io geth-devel: 	/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/eth/api_tracer.go:220 +0x698 

```

**Comment by gballet** [comment]
Er... you're welcome? :D

**Comment by gballet** [comment]
I could reproduce it with a `{}` object. I guess that I am simply missing some important value and that the de-serialization doesn't check for it.

**Comment by karalabe** [comment]
> ooooh ,I know
> I moved source code out of the trie ref counter
> originally a trie node could reference a) a node b) a subtrie c) source code
> the c) part really messes up pruning, so I just committed that directly to disk instead of refcounting
> but forgot tp update the tracer to do the same
> it messes it up because the pruner cannot decide if the arbitrary bytecode blob is a trie node or an evm blob
> so I would need to start special tracking for different data within the trie cache to keep the functionality... but a) code doesn't really weight much, contracts don't regularly self destruct, so it's not heavy; b) code cannot be pruned, because there's absolutely no way to know which accounts might reference the same code without duplicating it for all accounts
> tl;dr it's not possible to prune and it's not worth to bother... then less complexity is better
